equinox/nn/_embedding.py

```diff
@@ -217,35 +220,39 @@ def __call__(
             )
 
         with jax.ensure_compile_time_eval():
-            if embedding_size in internal_rope_embedding_cache:
-                freqs_cis = internal_rope_embedding_cache[embedding_size]
-                freqs_cis_seq_len, _ = freqs_cis.shape
-                if seq_len > freqs_cis_seq_len:
-                    freqs_cis = self.precompute_freqs_cis(
-                        embedding_size, seq_len, self.theta
-                    )
-                    internal_rope_embedding_cache[embedding_size] = freqs_cis
-                else:
-                    freqs_cis = freqs_cis[:seq_len]
-            else:
-                freqs_cis = self.precompute_freqs_cis(
-                    embedding_size, seq_len, self.theta
+            cache_key = (embedding_size, self.dtype)
+            if cache_key not in internal_rope_embedding_cache:
+                internal_rope_embedding_cache[cache_key] = self.precompute_freqs_cis(
+                    embedding_size, seq_len, self.theta, self.dtype
+                )
+
+            freqs_cos, freqs_sin = internal_rope_embedding_cache[cache_key]
+            freqs_seq_len, _ = freqs_cos.shape
+            if seq_len > freqs_seq_len:
+                internal_rope_embedding_cache[cache_key] = self.precompute_freqs_cis(
+                    embedding_size, seq_len, self.theta, self.dtype
                 )
-                internal_rope_embedding_cache[embedding_size] = freqs_cis
+                freqs_cos, freqs_sin = internal_rope_embedding_cache[cache_key]
+
+            freqs_cos = freqs_cos[:seq_len]
+            freqs_sin = freqs_sin[:seq_len]
 
-        freqs_real = jnp.tile(freqs_cis.real, (1, 2))
-        freqs_imag = jnp.tile(freqs_cis.imag, (1, 2))
+        freqs_cos = jnp.tile(freqs_cos, (1, 2))
+        freqs_sin = jnp.tile(freqs_sin, (1, 2))
 
         rotate_x = self.rotate_half(x)
-        x_rope = (x * freqs_real) + (rotate_x * freqs_imag)
-        return x_rope
+        x_rope = (x * freqs_cos) + (rotate_x * freqs_sin)
+        return x_rope.astype(x.dtype)
```

Comment:
I think this should be `x_rope = (x * freqs_cos.astype(x.dtype)) + (rotate_x * freqs_sin.astype(x.dtype))` ?

Right now you seem to be working around the lack of this with `with jax.numpy_dtype_promotion("standard"):` in the test.

Reply:
Thanks!